File: src/debloater.rs
//! Samsung device debloater for entry-level phones

use crate::adb::AdbManager;
use std::sync::Arc;

/// Debloat intensity level
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DebloatLevel {
    Light,      // Only obvious bloatware
    Medium,     // + AR, Games, Themes
    Aggressive, // + Knox, Health, Pay, most Samsung apps
    Custom,     // User-selected packages
}

impl DebloatLevel {
    pub fn description(&self) -> &'static str {
        match self {
            DebloatLevel::Light => "Safe - Remove obvious bloatware only",
            DebloatLevel::Medium => "Moderate - Remove AR, Games, Themes",
            DebloatLevel::Aggressive => "Maximum - Remove most Samsung apps",
            DebloatLevel::Custom => "Custom - Select packages manually",
        }
    }
}

/// Category of bloatware package
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DebloatCategory {
    Bixby,
    ArEmoji,
    Games,
    Themes,
    Knox,
    Health,
    Pay,
    Microsoft,
    Facebook,
    Google,
    SamsungApps,
    Carrier,
    Other,
}

impl DebloatCategory {
    pub fn name(&self) -> &'static str {
        match self {
            DebloatCategory::Bixby => "Bixby",
            DebloatCategory::ArEmoji => "AR Emoji",
            DebloatCategory::Games => "Games",
            DebloatCategory::Themes => "Themes",
            DebloatCategory::Knox => "Knox",
            DebloatCategory::Health => "Health",
            DebloatCategory::Pay => "Samsung Pay",
            DebloatCategory::Microsoft => "Microsoft",
            DebloatCategory::Facebook => "Facebook",
            DebloatCategory::Google => "Google",
            DebloatCategory::SamsungApps => "Samsung Apps",
            DebloatCategory::Carrier => "Carrier",
            DebloatCategory::Other => "Other",
        }
    }
}

/// A package that can be debloated
#[derive(Debug, Clone)]
pub struct DebloatPackage {
    pub package: String,
    pub name: String,
    pub category: DebloatCategory,
    pub description: String,
    pub is_selected: bool,
}

/// Samsung Debloater
pub struct Debloater {
    adb: Arc<AdbManager>,
}

impl Debloater {
    pub fn new(adb: Arc<AdbManager>) -> Self {
        Self { adb }
    }

    /// Check if device is a Samsung entry-level device
    pub fn is_entry_level_samsung(&self) -> bool {
        if let Ok(model) = self.adb.get_prop("ro.product.model") {
            let model = model.to_uppercase();
            // Entry-level A series
            model.contains("SM-A055") || // A05
            model.contains("SM-A065") || // A06
            model.contains("SM-A075") || // A07
            model.contains("SM-A165") || // A16
            model.contains("SM-A175") || // A17
            model.contains("SM-A135") || // A13
            model.contains("SM-A145") || // A14
            model.contains("SM-A155") || // A15
            model.contains("SM-A045") || // A04
            model.contains("SM-A046") || // A04s
            model.contains("SM-A035") || // A03
            model.contains("SM-A037") || // A03s
            model.contains("SM-A032") || // A03 Core
            // Entry-level M series
            model.contains("SM-M135") || // M13
            model.contains("SM-M145") || // M14
            model.contains("SM-M155")    // M15
        } else {
            false
        }
    }

    /// Get all debloatable packages with their categories
    pub fn get_all_packages() -> Vec<DebloatPackage> {
        let mut packages = Vec::new();

        // Bixby
        for (pkg, name, desc) in BIXBY_PACKAGES {
            packages.push(DebloatPackage {
                package: pkg.to_string(),
                name: name.to_string(),
                category: DebloatCategory::Bixby,
                description: desc.to_string(),
                is_selected: false,
            });
        }

        // AR Emoji
        for (pkg, name, desc) in AR_EMOJI_PACKAGES {
            packages.push(DebloatPackage {
                package: pkg.to_string(),
                name: name.to_string(),
                category: DebloatCategory::ArEmoji,
                description: desc.to_string(),
                is_selected: false,
            });
        }

        // Games
        for (pkg, name, desc) in GAMES_PACKAGES {
            packages.push(DebloatPackage {
                package: pkg.to_string(),
                name: name.to_string(),
                category: DebloatCategory::Games,
                description: desc.to_string(),
                is_selected: false,
            });
        }

        // Themes
        for (pkg, name, desc) in THEMES_PACKAGES {
            packages.push(DebloatPackage {
                package: pkg.to_string(),
                name: name.to_string(),
                category: DebloatCategory::Themes,
                description: desc.to_string(),
                is_selected: false,
            });
        }

        // Knox
        for (pkg, name, desc) in KNOX_PACKAGES {
            packages.push(DebloatPackage {
                package: pkg.to_string(),
                name: name.to_string(),
                category: DebloatCategory::Knox,
                description: desc.to_string(),
                is_selected: false,
            });
        }

        // Health
        for (pkg, name, desc) in HEALTH_PACKAGES {
            packages.push(DebloatPackage {
                package: pkg.to_string(),
                name: name.to_string(),
                category: DebloatCategory::Health,
                description: desc.to_string(),
                is_selected: false,
            });
        }

        // Samsung Pay
        for (pkg, name, desc) in PAY_PACKAGES {
            packages.push(DebloatPackage {
                package: pkg.to_string(),
                name: name.to_string(),
                category: DebloatCategory::Pay,
                description: desc.to_string(),
                is_selected: false,
            });
        }

        // Microsoft
        for (pkg, name, desc) in MICROSOFT_PACKAGES {
            packages.push(DebloatPackage {
                package: pkg.to_string(),
                name: name.to_string(),
                category: DebloatCategory::Microsoft,
                description: desc.to_string(),
                is_selected: false,
            });
        }

        // Facebook
        for (pkg, name, desc) in FACEBOOK_PACKAGES {
            packages.push(DebloatPackage {
                package: pkg.to_string(),
                name: name.to_string(),
                category: DebloatCategory::Facebook,
                description: desc.to_string(),
                is_selected: false,
            });
        }

        // Samsung Apps
        for (pkg, name, desc) in SAMSUNG_APPS_PACKAGES {
            packages.push(DebloatPackage {
                package: pkg.to_string(),
                name: name.to_string(),
                category: DebloatCategory::SamsungApps,
                description: desc.to_string(),
                is_selected: false,
            });
        }

        packages
    }

    /// Get packages for a specific debloat level
    pub fn get_packages_for_level(level: DebloatLevel) -> Vec<DebloatPackage> {
        let all = Self::get_all_packages();

        match level {
            DebloatLevel::Light => {
                all.into_iter()
                    .filter(|p| matches!(p.category,
                        DebloatCategory::Facebook |
                        DebloatCategory::Games))
                    .collect()
            }
            DebloatLevel::Medium => {
                all.into_iter()
                    .filter(|p| matches!(p.category,
                        DebloatCategory::Facebook |
                        DebloatCategory::Games |
                        DebloatCategory::ArEmoji |
                        DebloatCategory::Themes |
                        DebloatCategory::Bixby))
                    .collect()
            }
            DebloatLevel::Aggressive => all,
            DebloatLevel::Custom => all,
        }
    }

    /// Debloat a single package (uninstall for user 0)
    pub fn debloat_package(&self, package: &str) -> anyhow::Result<bool> {
        // Try uninstall first
        if self.adb.uninstall_package(package)? {
            return Ok(true);
        }

        // Fall back to disable
        self.adb.disable_package(package)
    }

    /// Debloat multiple packages
    pub fn debloat_packages(&self, packages: &[String]) -> Vec<(String, bool)> {
        packages.iter()
            .map(|pkg| {
                let success = self.debloat_package(pkg).unwrap_or(false);
                (pkg.clone(), success)
            })
            .collect()
    }

    /// Get which packages from the list are actually installed
    pub fn get_installed_packages(&self, packages: &[DebloatPackage]) -> anyhow::Result<Vec<DebloatPackage>> {
        let installed = self.adb.get_packages(true)?;
        let installed_names: std::collections::HashSet<_> = installed
            .iter()
            .map(|p| p.package_name.as_str())
            .collect();

        Ok(packages.iter()
            .filter(|p| installed_names.contains(p.package.as_str()))
            .cloned()
            .collect())
    }
}

// Package lists
const BIXBY_PACKAGES: &[(&str, &str, &str)] = &[
    ("com.samsung.android.bixby.agent", "Bixby Voice", "Voice assistant"),
    ("com.samsung.android.bixby.service", "Bixby Service", "Background service"),
    ("com.samsung.android.visionintelligence", "Bixby Vision", "Camera AI features"),
    ("com.samsung.android.bixby.wakeup", "Bixby Wakeup", "Voice activation"),
    ("com.samsung.android.app.spage", "Samsung Free", "News/content feed"),
    ("com.samsung.android.app.routines", "Bixby Routines", "Automation"),
    ("com.samsung.android.bixbyvision.framework", "Bixby Vision Framework", "Vision core"),
];

const AR_EMOJI_PACKAGES: &[(&str, &str, &str)] = &[
    ("com.samsung.android.aremoji", "AR Emoji", "Create AR emoji"),
    ("com.samsung.android.arzone", "AR Zone", "AR features hub"),
    ("com.samsung.android.ardrawing", "AR Doodle", "Draw in AR"),
    ("com.samsung.android.aremojieditor", "AR Emoji Editor", "Edit emoji"),
    ("com.sec.android.mimage.avatarstickers", "Avatar Stickers", "Emoji stickers"),
    ("com.samsung.android.app.camera.sticker.facearavatar.preload", "AR Stickers", "Camera stickers"),
];

const GAMES_PACKAGES: &[(&str, &str, &str)] = &[
    ("com.samsung.android.game.gamehome", "Game Home", "Game launcher"),
    ("com.samsung.android.game.gametools", "Game Tools", "Game overlay"),
    ("com.samsung.android.game.gos", "Game Optimizing", "Game performance"),
    ("com.enhance.gameservice", "Game Service", "Gaming features"),
    ("com.samsung.android.game.gameService", "Game Service", "Gaming backend"),
];

const THEMES_PACKAGES: &[(&str, &str, &str)] = &[
    ("com.samsung.android.themestore", "Galaxy Themes", "Theme store"),
    ("com.samsung.android.themecenter", "Theme Center", "Theme manager"),
    ("com.samsung.android.app.dressroom", "Wallpapers", "Wallpaper app"),
];

const KNOX_PACKAGES: &[(&str, &str, &str)] = &[
    ("com.samsung.android.knox.containercore", "Knox Container", "Secure folder core"),
    ("com.samsung.android.knox.analytics.uploader", "Knox Analytics", "Enterprise analytics"),
    ("com.samsung.android.knox.pushmanager", "Knox Push", "Enterprise push"),
    ("com.samsung.android.knox.kpu", "Knox Platform", "Enterprise platform"),
    ("com.sec.enterprise.knox.attestation", "Knox Attestation", "Security verification"),
    ("com.samsung.knox.securefolder", "Secure Folder", "Private space"),
    ("com.samsung.android.appseparation", "App Separation", "Knox isolation"),
];

const HEALTH_PACKAGES: &[(&str, &str, &str)] = &[
    ("com.samsung.android.forest", "Digital Wellbeing", "Screen time"),
    ("com.samsung.android.app.parentalcare", "Parental Controls", "Family features"),
    ("com.sec.android.app.shealth", "Samsung Health", "Health tracking"),
    ("com.samsung.android.service.health", "Health Service", "Health backend"),
];

const PAY_PACKAGES: &[(&str, &str, &str)] = &[
    ("com.samsung.android.spay", "Samsung Pay", "Mobile payments"),
    ("com.samsung.android.spayfw", "Samsung Pay Framework", "Payment core"),
    ("com.samsung.android.samsungpass", "Samsung Pass", "Password manager"),
    ("com.samsung.android.samsungpassautofill", "Pass Autofill", "Autofill service"),
    ("com.samsung.android.authfw", "Auth Framework", "Authentication"),
];

const MICROSOFT_PACKAGES: &[(&str, &str, &str)] = &[
    ("com.microsoft.appmanager", "MS App Manager", "Microsoft installer"),
    ("com.microsoft.office.outlook", "Outlook", "Email client"),
    ("com.microsoft.office.officehubrow", "Office Hub", "Office suite"),
    ("com.microsoft.skydrive", "OneDrive", "Cloud storage"),
    ("com.linkedin.android", "LinkedIn", "Professional network"),
    ("com.microsoft.office.word", "Word", "Document editor"),
    ("com.microsoft.office.excel", "Excel", "Spreadsheets"),
    ("com.microsoft.office.powerpoint", "PowerPoint", "Presentations"),
];

const FACEBOOK_PACKAGES: &[(&str, &str, &str)] = &[
    ("com.facebook.katana", "Facebook", "Social network"),
    ("com.facebook.orca", "Messenger", "Chat app"),
    ("com.facebook.system", "Facebook Services", "Background service"),
    ("com.facebook.appmanager", "FB App Manager", "Facebook installer"),
    ("com.facebook.services", "FB Services", "Facebook backend"),
    ("com.instagram.android", "Instagram", "Photo sharing"),
];

const SAMSUNG_APPS_PACKAGES: &[(&str, &str, &str)] = &[
    ("com.sec.android.app.sbrowser", "Samsung Browser", "Web browser"),
    ("com.samsung.android.app.tips", "Samsung Tips", "Tips and tricks"),
    ("com.samsung.android.mobileservice", "Samsung Experience", "Samsung services"),
    ("com.samsung.android.app.watchmanagerstub", "Galaxy Wearable", "Watch app stub"),
    ("com.samsung.android.voc", "Samsung Members", "Support app"),
    ("com.samsung.android.app.social", "Samsung Social", "What's New feed"),
    ("com.samsung.android.app.reminder", "Reminder", "Tasks app"),
    ("com.samsung.android.calendar", "Samsung Calendar", "Calendar app"),
    ("com.samsung.android.email.provider", "Email", "Email client"),
    ("com.samsung.android.app.notes", "Samsung Notes", "Note taking"),
    ("com.samsung.android.svoiceime", "Samsung Voice Input", "Voice typing"),
    ("com.samsung.android.kidsinstaller", "Kids Mode", "Parental mode"),
    ("com.samsung.android.app.sharelive", "Quick Share", "File sharing"),
    ("com.samsung.android.smartsuggestions", "Smart Suggestions", "AI suggestions"),
    ("com.samsung.android.privateshare", "Private Share", "Secure sharing"),
];

File: AndroidSuperTool/Managers/Debloater.swift
//
//  Debloater.swift
//  AndroidSuperTool
//

import Foundation

/// Debloat intensity level
enum DebloatLevel: String, CaseIterable {
    case light = "Light"
    case medium = "Medium"
    case aggressive = "Aggressive"
    case custom = "Custom"

    var description: String {
        switch self {
        case .light: return "Safe - Remove obvious bloatware only"
        case .medium: return "Moderate - Remove AR, Games, Themes"
        case .aggressive: return "Maximum - Remove most Samsung apps"
        case .custom: return "Custom - Select packages manually"
        }
    }
}

/// Category of bloatware package
enum DebloatCategory: String, CaseIterable {
    case bixby = "Bixby"
    case arEmoji = "AR Emoji"
    case games = "Games"
    case themes = "Themes"
    case knox = "Knox"
    case health = "Health"
    case pay = "Samsung Pay"
    case microsoft = "Microsoft"
    case facebook = "Facebook"
    case samsungApps = "Samsung Apps"
}

/// A package that can be debloated
class DebloatPackage: Identifiable, ObservableObject {
    let id = UUID()
    let packageName: String
    let name: String
    let category: DebloatCategory
    let description: String
    @Published var isSelected: Bool = false

    init(packageName: String, name: String, category: DebloatCategory, description: String) {
        self.packageName = packageName
        self.name = name
        self.category = category
        self.description = description
    }
}

/// Samsung Debloater
class Debloater {
    private let adb: ADBManager

    init(adb: ADBManager) {
        self.adb = adb
    }

    /// Check if device is a Samsung entry-level device
    func isEntryLevelSamsung() -> Bool {
        guard let model = adb.getProperty("ro.product.model")?.uppercased() else {
            return false
        }

        let entryLevelPrefixes = [
            "SM-A055", "SM-A065", "SM-A075", "SM-A165", "SM-A175",
            "SM-A135", "SM-A145", "SM-A155", "SM-A045", "SM-A046",
            "SM-A035", "SM-A037", "SM-A032", "SM-M135", "SM-M145", "SM-M155"
        ]

        return entryLevelPrefixes.contains { model.hasPrefix($0) }
    }

    /// Get all debloatable packages
    static func getAllPackages() -> [DebloatPackage] {
        var packages = [DebloatPackage]()

        // Bixby
        for (pkg, name, desc) in bixbyPackages {
            packages.append(DebloatPackage(packageName: pkg, name: name, category: .bixby, description: desc))
        }

        // AR Emoji
        for (pkg, name, desc) in arEmojiPackages {
            packages.append(DebloatPackage(packageName: pkg, name: name, category: .arEmoji, description: desc))
        }

        // Games
        for (pkg, name, desc) in gamesPackages {
            packages.append(DebloatPackage(packageName: pkg, name: name, category: .games, description: desc))
        }

        // Themes
        for (pkg, name, desc) in themesPackages {
            packages.append(DebloatPackage(packageName: pkg, name: name, category: .themes, description: desc))
        }

        // Knox
        for (pkg, name, desc) in knoxPackages {
            packages.append(DebloatPackage(packageName: pkg, name: name, category: .knox, description: desc))
        }

        // Health
        for (pkg, name, desc) in healthPackages {
            packages.append(DebloatPackage(packageName: pkg, name: name, category: .health, description: desc))
        }

        // Samsung Pay
        for (pkg, name, desc) in payPackages {
            packages.append(DebloatPackage(packageName: pkg, name: name, category: .pay, description: desc))
        }

        // Microsoft
        for (pkg, name, desc) in microsoftPackages {
            packages.append(DebloatPackage(packageName: pkg, name: name, category: .microsoft, description: desc))
        }

        // Facebook
        for (pkg, name, desc) in facebookPackages {
            packages.append(DebloatPackage(packageName: pkg, name: name, category: .facebook, description: desc))
        }

        // Samsung Apps
        for (pkg, name, desc) in samsungAppsPackages {
            packages.append(DebloatPackage(packageName: pkg, name: name, category: .samsungApps, description: desc))
        }

        return packages
    }

    /// Get packages for a specific debloat level
    static func getPackagesForLevel(_ level: DebloatLevel) -> [DebloatPackage] {
        let all = getAllPackages()

        switch level {
        case .light:
            return all.filter { $0.category == .facebook || $0.category == .games }
        case .medium:
            return all.filter {
                [.facebook, .games, .arEmoji, .themes, .bixby].contains($0.category)
            }
        case .aggressive, .custom:
            return all
        }
    }

    /// Debloat a single package
    @discardableResult
    func debloatPackage(_ packageName: String) -> Bool {
        // Try uninstall first
        if adb.uninstallPackage(packageName) {
            return true
        }

        // Fall back to disable
        return adb.disablePackage(packageName)
    }

    /// Debloat multiple packages
    func debloatPackages(_ packages: [String]) -> [(String, Bool)] {
        packages.map { ($0, debloatPackage($0)) }
    }

    /// Get which packages from the list are actually installed
    func getInstalledPackages(_ packages: [DebloatPackage]) -> [DebloatPackage] {
        let installed = Set(adb.getPackages(includeSystem: true).map { $0.packageName })
        return packages.filter { installed.contains($0.packageName) }
    }

    // Package lists
    private static let bixbyPackages: [(String, String, String)] = [
        ("com.samsung.android.bixby.agent", "Bixby Voice", "Voice assistant"),
        ("com.samsung.android.bixby.service", "Bixby Service", "Background service"),
        ("com.samsung.android.visionintelligence", "Bixby Vision", "Camera AI features"),
        ("com.samsung.android.bixby.wakeup", "Bixby Wakeup", "Voice activation"),
        ("com.samsung.android.app.spage", "Samsung Free", "News/content feed"),
        ("com.samsung.android.app.routines", "Bixby Routines", "Automation"),
    ]

    private static let arEmojiPackages: [(String, String, String)] = [
        ("com.samsung.android.aremoji", "AR Emoji", "Create AR emoji"),
        ("com.samsung.android.arzone", "AR Zone", "AR features hub"),
        ("com.samsung.android.ardrawing", "AR Doodle", "Draw in AR"),
        ("com.samsung.android.aremojieditor", "AR Emoji Editor", "Edit emoji"),
        ("com.sec.android.mimage.avatarstickers", "Avatar Stickers", "Emoji stickers"),
    ]

    private static let gamesPackages: [(String, String, String)] = [
        ("com.samsung.android.game.gamehome", "Game Home", "Game launcher"),
        ("com.samsung.android.game.gametools", "Game Tools", "Game overlay"),
        ("com.samsung.android.game.gos", "Game Optimizing", "Game performance"),
        ("com.enhance.gameservice", "Game Service", "Gaming features"),
    ]

    private static let themesPackages: [(String, String, String)] = [
        ("com.samsung.android.themestore", "Galaxy Themes", "Theme store"),
        ("com.samsung.android.themecenter", "Theme Center", "Theme manager"),
        ("com.samsung.android.app.dressroom", "Wallpapers", "Wallpaper app"),
    ]

    private static let knoxPackages: [(String, String, String)] = [
        ("com.samsung.android.knox.containercore", "Knox Container", "Secure folder core"),
        ("com.samsung.android.knox.analytics.uploader", "Knox Analytics", "Enterprise analytics"),
        ("com.samsung.android.knox.pushmanager", "Knox Push", "Enterprise push"),
        ("com.samsung.knox.securefolder", "Secure Folder", "Private space"),
    ]

    private static let healthPackages: [(String, String, String)] = [
        ("com.samsung.android.forest", "Digital Wellbeing", "Screen time"),
        ("com.samsung.android.app.parentalcare", "Parental Controls", "Family features"),
        ("com.sec.android.app.shealth", "Samsung Health", "Health tracking"),
    ]

    private static let payPackages: [(String, String, String)] = [
        ("com.samsung.android.spay", "Samsung Pay", "Mobile payments"),
        ("com.samsung.android.spayfw", "Samsung Pay Framework", "Payment core"),
        ("com.samsung.android.samsungpass", "Samsung Pass", "Password manager"),
    ]

    private static let microsoftPackages: [(String, String, String)] = [
        ("com.microsoft.appmanager", "MS App Manager", "Microsoft installer"),
        ("com.microsoft.office.outlook", "Outlook", "Email client"),
        ("com.microsoft.office.officehubrow", "Office Hub", "Office suite"),
        ("com.microsoft.skydrive", "OneDrive", "Cloud storage"),
        ("com.linkedin.android", "LinkedIn", "Professional network"),
    ]

    private static let facebookPackages: [(String, String, String)] = [
        ("com.facebook.katana", "Facebook", "Social network"),
        ("com.facebook.orca", "Messenger", "Chat app"),
        ("com.facebook.system", "Facebook Services", "Background service"),
        ("com.facebook.appmanager", "FB App Manager", "Facebook installer"),
        ("com.instagram.android", "Instagram", "Photo sharing"),
    ]

    private static let samsungAppsPackages: [(String, String, String)] = [
        ("com.sec.android.app.sbrowser", "Samsung Browser", "Web browser"),
        ("com.samsung.android.app.tips", "Samsung Tips", "Tips and tricks"),
        ("com.samsung.android.mobileservice", "Samsung Experience", "Samsung services"),
        ("com.samsung.android.voc", "Samsung Members", "Support app"),
        ("com.samsung.android.calendar", "Samsung Calendar", "Calendar app"),
        ("com.samsung.android.app.notes", "Samsung Notes", "Note taking"),
    ]
}
